Sorry about all the pushes.

I have a NixOS test now that WorksForMe:tm: on my x86_64-linux machine and on x86_64-linux CI, but aarch64-linux CI seems to struggle with it.

<details>
  <summary>Opening the first Wayland-native terminal via the window manager's keybind is consistently not happening. In the logs I see that the keybind was fired, but I don't see the startup logs of the terminal itself, and it times out waiting for the OCR to see the terminal's prompt.</summary>

```
(finished: sending monitor command: sendkey ctrl-alt-t, in 0.00 seconds)
machine: waiting for alice@machine to appear on screen
# ...
error: building of '/nix/store/48njg2m0gkydd3phfh339a333vljhg8n-vm-test-run-miriway.drv!out' from .drv file timed out after 3600 seconds
```
</details>

<details>
  <summary>I've made it so it opens the terminal on startup, in case it was trying to fire the keybind before Miriway can process it. That works, and it tries to run a first test command to check the Wayland environment. The command should generate a file with the logs, but it looks like it's stuck waiting for that file now.</summary>

```
(finished: waiting for alice@machine to appear on screen, in 12.93 seconds)
machine: sending keys 'test-wayland
'
# ...
(finished: sending keys 'test-wayland
', in 0.13 seconds)
machine: waiting for file '/tmp/test-wayland-exit-ok'
# ...
cleanup
kill machine (pid 8)
machine # qemu-system-aarch64: terminating on signal 15 from pid 6 (/nix/store/il2zyns5nlpi53sczvgg3mdpkm5z524z-python3-3.10.9/bin/python3.10)
(finished: cleanup, in 0.05 seconds)
Traceback (most recent call last):
  File "/nix/store/vigv5wxf1y4qkclwg1hrjivl7br0w5hs-nixos-test-driver-1.1/bin/.nixos-test-driver-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/vigv5wxf1y4qkclwg1hrjivl7br0w5hs-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/__init__.py", line 112, in main
    driver.run_tests()
  File "/nix/store/vigv5wxf1y4qkclwg1hrjivl7br0w5hs-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 142, in run_tests
    self.test_script()
  File "/nix/store/vigv5wxf1y4qkclwg1hrjivl7br0w5hs-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 138, in test_script
    exec(self.tests, symbols, None)
  File "<string>", line 14, in <module>
  File "/nix/store/vigv5wxf1y4qkclwg1hrjivl7br0w5hs-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 700, in wait_for_file
    retry(check_file)
  File "/nix/store/vigv5wxf1y4qkclwg1hrjivl7br0w5hs-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 131, in retry
    raise Exception(f"action timed out after {timeout} seconds")
Exception: action timed out after 900 seconds
```
</details>

<details>
  <summary>It looks like it fails to receive the keyboard inputs in general? On x86_64-linux I see that Mir is recognising & registering new evdev input devices when the first keyboard inputs are sent. I checked that it is indeed connected to the inputs getting sent and not a timing thing by making it sleep 2, 3, 5 seconds before sending the keybind.<br/>
On aarch64-linux CI, no such output is sent. Maybe this is a problem with how the VM inputs are configured? In which case I don't think I'm NixOS test / VM savvy enough to know how to address this.</summary>

```
machine # [   22.696751] xsession[870]: [2023-02-03 19:14:40.580481] <information> evdev-input: Added /dev/input/event1: QEMU USB Tablet
machine # [   22.698601] xsession[870]: [2023-02-03 19:14:40.580709] <information> evdev-input: Added /dev/input/event3
machine # [   22.709933] xsession[870]: [2023-02-03 19:14:40.597750] <information> evdev-input: Added /dev/input/event0
machine # [   22.723355] xsession[870]: [2023-02-03 19:14:40.611172] <information> evdev-input: Added /dev/input/event5
machine # [   22.724655] xsession[870]: [2023-02-03 19:14:40.612472] <information> evdev-input: Added /dev/input/event4
```
</details>

I'll check if I can replicate this on my aarch64-linux hardware and try to debug & fix it there, though I'm not sure *how* exactly :sweat:.

--- a/nixos/tests/all-tests.nix
+++ b/nixos/tests/all-tests.nix
@@ -395,6 +395,7 @@ in {
   minidlna = handleTest ./minidlna.nix {};
   miniflux = handleTest ./miniflux.nix {};
   minio = handleTest ./minio.nix {};
+  miriway = handleTest ./miriway.nix {};
   misc = handleTest ./misc.nix {};
   mjolnir = handleTest ./matrix/mjolnir.nix {};
   mod_perl = handleTest ./mod_perl.nix {};

--- a/nixos/tests/miriway.nix
+++ b/nixos/tests/miriway.nix
@@ -1,0 +1,126 @@
+import ./make-test-python.nix ({ pkgs, lib, ... }: {
+  name = "miriway";
+
+  meta = {
+    maintainers = with lib.maintainers; [ OPNA2608 ];
+    # FIXME On ARM Miriway inside the VM doesn't receive keyboard inputs, why?
+    broken = pkgs.stdenv.hostPlatform.isAarch;
+  };
+
+  nodes.machine = { config, ... }: {
+    imports = [
+      ./common/auto.nix
+      ./common/user-account.nix
+    ];
+
+    # Seems to very rarely get interrupted by oom-killer
+    virtualisation.memorySize = 2047;
+
+    test-support.displayManager.auto = {
+      enable = true;
+      user = "alice";
+    };
+
+    services.xserver = {
+      enable = true;
+      displayManager.defaultSession = lib.mkForce "miriway";
+    };
+
+    programs.miriway = {
+      enable = true;
+      config = ''
+        add-wayland-extensions=all
+
+        ctrl-alt=t:foot --maximized
+        ctrl-alt=a:env WINIT_UNIX_BACKEND=x11 WAYLAND_DISPLAY=invalid alacritty --option window.startup_mode=maximized
+
+        shell-component=dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY
+
+        shell-component=foot --maximized
+      '';
+    };
+
+    environment = {
+      shellAliases = {
+        test-wayland = "wayland-info | tee /tmp/test-wayland.out && touch /tmp/test-wayland-exit-ok";
+        test-x11 = "glinfo | tee /tmp/test-x11.out && touch /tmp/test-x11-exit-ok";
+      };
+
+      systemPackages = with pkgs; [
+        mesa-demos
+        wayland-utils
+        foot
+        alacritty
+      ];
+
+      # To help with OCR
+      etc."xdg/foot/foot.ini".text = lib.generators.toINI { } {
+        main = {
+          font = "inconsolata:size=16";
+        };
+        colors = rec {
+          foreground = "000000";
+          background = "ffffff";
+          regular2 = foreground;
+        };
+      };
+      etc."xdg/alacritty/alacritty.yml".text = lib.generators.toYAML { } {
+        font = rec {
+          normal.family = "Inconsolata";
+          bold.family = normal.family;
+          italic.family = normal.family;
+          bold_italic.family = normal.family;
+          size = 16;
+        };
+        colors = rec {
+          primary = {
+            foreground = "0x000000";
+            background = "0xffffff";
+          };
+          normal = {
+            green = primary.foreground;
+          };
+        };
+      };
+    };
+
+    fonts.fonts = [ pkgs.inconsolata ];
+  };
+
+  enableOCR = true;
+
+  testScript = { nodes, ... }: ''
+    start_all()
+    machine.wait_for_unit("multi-user.target")
+
+    # Wait for Miriway to complete startup
+    machine.wait_for_file("/run/user/1000/wayland-0")
+    machine.succeed("pgrep miriway-shell")
+    machine.screenshot("miriway_launched")
+
+    # Test Wayland
+    # We let Miriway start the first terminal, as we might get stuck if it's not ready to process the first keybind
+    # machine.send_key("ctrl-alt-t")
+    machine.wait_for_text("alice@machine")
+    machine.send_chars("test-wayland\n")
+    machine.wait_for_file("/tmp/test-wayland-exit-ok")
+    machine.copy_from_vm("/tmp/test-wayland.out")
+    machine.screenshot("foot_wayland_info")
+    # Only succeeds when a mouse is moved inside an interactive session?
+    # machine.send_chars("exit\n")
+    # machine.wait_until_fails("pgrep foot")
+    machine.succeed("pkill foot")
+
+    # Test XWayland
+    machine.send_key("ctrl-alt-a")
+    machine.wait_for_text("alice@machine")
+    machine.send_chars("test-x11\n")
+    machine.wait_for_file("/tmp/test-x11-exit-ok")
+    machine.copy_from_vm("/tmp/test-x11.out")
+    machine.screenshot("alacritty_glinfo")
+    # Only succeeds when a mouse is moved inside an interactive session?
+    # machine.send_chars("exit\n")
+    # machine.wait_until_fails("pgrep alacritty")
+    machine.succeed("pkill alacritty")
+  '';
+})

--- a/pkgs/applications/window-managers/miriway/default.nix
+++ b/pkgs/applications/window-managers/miriway/default.nix
@@ -2,6 +2,7 @@
 , lib
 , fetchFromGitHub
 , unstableGitUpdater
+, nixosTests
 , cmake
 , pkg-config
 , mir
@@ -43,6 +44,9 @@ stdenv.mkDerivation rec {
   passthru = {
     updateScript = unstableGitUpdater { };
     providedSessions = [ "miriway" ];
+    tests = {
+      inherit (nixosTests) miriway;
+    };
   };
 
   meta = with lib; {

